src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -268,18 +358,159 @@ public void WriteValue(byte[] value)
+        /// This method should not be called by <see cref="WriteRawValue(string)"/>.
+        /// </summary>
+        private void WriteSeparatorIfNecessary()
+        {
+            if (this.shouldWriteSeparator)
+            {
+                this.WriteItemSeparator();
+            }
+
+            // reset state since now we're back in sync with Utf8JsonWriter.
+            this.shouldWriteSeparator = false;
+            this.isWritingFirstElementInArray = false;
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+        }
+
+        /// <summary>
+        /// Manually writes an item separator ',' to the output, bypassing the Utf8JsonWriter.
+        /// </summary>
+        private void WriteItemSeparator()
+        {
+            this.CommitWriterContentsToBuffer();
+            Span<byte> buf = this.bufferWriter.GetSpan(1);
+            buf[0] = (byte)',';
+            this.bufferWriter.Advance(1);
+        }
+
+        /// <summary>
+        /// Mark that the writer has transitioned to write an object.
+        /// </summary>
+        private void EnterObjectScope()
+        {
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+            this.isWritingFirstElementInArray = false;
+            this.bitStack.PushTrue();
+        }
+
+        /// <summary>
+        /// Mark that the writer has transitioned to write an array.
+        /// </summary>
+        private void EnterArrayScope()
+        {
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+            this.isWritingFirstElementInArray = true;
+            this.bitStack.PushFalse();
+        }
+
+        /// <summary>
+        /// Exit the current object or array scope.
+        /// </summary>
+        private void ExitScope()
+        {
+            this.isWritingFirstElementInArray = false;
+            this.shouldWriteSeparator = false;
+            this.bitStack.Pop();
+        }
+
+        /// <summary>
+        /// Checks whether the writer is currently in an array scope.
+        /// </summary>
+        /// <returns>true if the writer is currently in an array scope; otherwise false.</returns>
+        private bool IsInArray()
+        {
+            if (!TryPeekScope(out bool isInObject))
+            {
+                return false;
+            }
+
+            return !isInObject;
+        }
+
+        /// <summary>
+        /// Tries to get the current scope.
+        /// </summary>
+        /// <param name="isInObject">
+        /// <returns>True if the current scope was successfully determined, false if it was not possible to determine the scope
+        /// (i.e. we're neither in an array or object).
+        /// </returns>
+        private bool TryPeekScope(out bool isInObject)
+        {
+            isInObject = false;
+            if (this.bitStack.CurrentDepth == 0)
+            {
+                return false;
+            }
+
+            // BitStack doesn't implement a Peek()

```

Comment:
It would probably be more efficient to have a specific `Peek()` but it would increase the complexity of the code, especially if the depth > 64 (this is rare in practice, but we'd still need to implement the logic to handle it). Since `Peek()` and `Pop()` are relatively cheap and are aggressively inlined, I don't think we'll save that much in implementing a `Peek()` method. That's why I felt like it wasn't worth the effort. That said, I haven't actually measured the cost savings of implementing a `Peek()` (because doing so would require me to implement `Peek()` in the first place).